pkg/cmd/server/server.go

```diff
@@ -227,12 +236,25 @@ func (s *server) watchConfig(config *api.Config) {
 	s.sharedInformerFactory.Ark().V1().Configs().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		UpdateFunc: func(oldObj, newObj interface{}) {
 			updated := newObj.(*api.Config)
+			glog.V(4).Infof("received updated config: %q", kube.NamespaceAndName(updated))
 
 			if updated.Name != config.Name {
 				glog.V(5).Infof("config watch channel received other config %q", updated.Name)
 				return
 			}
 
+			// Objects retrieved via Get() don't have their Kind or APIVersion set. Objects retrieved via
+			// Watch(), including those from shared informer event handlers, DO have their Kind and
+			// APIVersion set. To prevent the DeepEqual() call below from considering Kind or APIVersion
+			// as the source of a change, set config.Kind and config.APIVersion to match the values from
+			// the updated Config.
```

Comment:
Is this a known issue with CRDs in general or something specific to our implementation?  

And a more general question around the `APIVersion`, would that be used to signify a major upgrade in our configuration and should that be a consideration to simply ignoring it?

Reply:
> Is this a known issue with CRDs in general or something specific to our implementation?

This is actually the way that the Kubernetes API machinery works. It has nothing to do with CRDs specifically, as you'll see the same behavior with the client-go clientset and Namespaces, for example.

> And a more general question around the APIVersion, would that be used to signify a major upgrade in our configuration and should that be a consideration to simply ignoring it?

In general, a stable non-alpha non-beta API should always be backwards compatible as the software evolves. Breaking changes are only allowed in a newer version. This is the case with e.g. `storage.k8s.io/v1` - you can't remove fields or change the types of existing fields. Within an API group (`storage.k8s.io`), you can have multiple versions.

And here is where we get to something that is specific to CRDs: the Kubernetes apiserver only supports a single APIVersion per API group. I could go into more detail as to why, but I'll save that for another time, and simply say that we are "stuck" on `ark.heptio.com/v1`, and it's up to us to decide how and when we want to provide compatibility guarantees. For right now we're saying we're alpha, and we may change the API types, and you may have to wipe your existing data and start over. We'll try not to break things, but we may have to.

tl;dr: we'll only ever get Kind=Config and APIVersion=ark.heptio.com/v1, so it's safe to do this hack.

Reply:
Thanks for taking the time with a detailed explanation, that's helpful context.